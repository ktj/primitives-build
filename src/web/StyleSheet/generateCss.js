const normalizeValue = require('./normalizeValue');
const hyphenate = require('./hyphenate');
const mapKeyValue = require('../util/mapKeyValue');
const transformToWebStyle = require('./transformToWebStyle');

/**
 * Generates valid CSS rule bodies (sans selector) from a JS object
 *
 * Example:
 * ```
 * generateCss({ width: 20, color: 'blue' });
 * //=> 'width:20px;color:blue;'
 * ```
 *
 * @param style
 */
const generateCss = style => mapKeyValue(transformToWebStyle(style), (key, val) => {
  const name = hyphenate(key);
  const value = normalizeValue(key, val);
  if (Array.isArray(val)) {
    return val.map(v => `${name}:${v};`).join('');
  }
  return `${name}:${value};`;
}).join('');

module.exports = generateCss;
