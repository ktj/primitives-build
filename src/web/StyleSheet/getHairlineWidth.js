// Modeled after: http://dieulot.net/css-retina-hairline

// NOTE:
// This assumes this script runs after <body>. If it is run inside of `<head>`, it will just
// return 1.
module.exports = () => {
  let hairlineWidth = 1;
  if (!global || !global.document || !global.document.body) {
    return hairlineWidth;
  }
  if (global.devicePixelRatio && global.devicePixelRatio >= 2) {
    const testElem = document.createElement('div');
    testElem.style.border = '.5px solid transparent';
    document.body.appendChild(testElem);
    if (testElem.offsetHeight === 1) {
      hairlineWidth = 0.5;
    }
    document.body.removeChild(testElem);
  }
  return hairlineWidth;
};
