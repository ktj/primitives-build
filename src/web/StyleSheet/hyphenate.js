// Taken from:
// https://github.com/necolas/react-native-web/blob/master/src/apis/StyleSheet/hyphenate.js

module.exports = s => s.replace(/([A-Z])/g, '-$1').toLowerCase().replace(/^ms-/, '-ms-');
