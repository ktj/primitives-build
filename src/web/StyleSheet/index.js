const getHairlineWidth = require('./getHairlineWidth');
const transformToWebStyle = require('./transformToWebStyle');
const { reset, registerStyle, getStyle, getClassNames } = require('./registry');
const init = require('./initialStyles');
const injector = require('./injector');

// TODO:
// 3. ensure that multiple-bundled react-primitives will work...
// 6. use invariant and add a lot of validation + runtime checks
// 7. introduce a "strict mode" which only allows RN-compatible API (not sure about this any more)
// 12. how to handle custom font families??

const hasOwnProperty = Object.prototype.hasOwnProperty;

const create = styles => {
  const result = {};
  Object.keys(styles).forEach(key => {
    result[key] = registerStyle(key, styles[key]);
  });
  return result;
};

const mergeTransforms = (a, b) => {
  if (!a || a.length === 0) return b; // in this case, a has nothing to contribute.
  const result = [];
  const transformsInA = a.reduce((hash, t) => {
    const key = Object.keys(t)[0];
    result.push(t);
    hash[key] = result.length - 1;
    return hash;
  }, {});
  b.forEach(t => {
    const key = Object.keys(t)[0];
    const index = transformsInA[key];
    if (index !== undefined) {
      result[index] = t;
    } else {
      result.push(t);
    }
  });
  return result;
};

// merge two style hashes together. Sort of like `Object.assign`, but is aware of `transform` as a
// special case.
// NOTE: mutates the first argument!
const assignStyle = (a, b) => {
  let key;
  for (key in b) {
    if (hasOwnProperty.call(b, key)) {
      switch (key) {
        case 'transform':
          a[key] = mergeTransforms(a[key], b[key]);
          break;
        default:
          /* eslint no-param-reassign: 0 */
          a[key] = b[key];
          break;
      }
    }
  }
  return a;
};

const flattenStyle = (input, expandRegisteredStyles) => {
  // optimize for this (common) scenario of input not needing flattening
  if (!Array.isArray(input)) {
    if (typeof input === 'number') {
      return expandRegisteredStyles ? getStyle(input) : undefined;
    }
    if (!input) {
      return undefined;
    }
    return input;
  }
  const result = {};
  let hasResult = false;
  const stack = [{ i: 0, array: input }];
  while (stack.length) {
    const iterator = stack.pop();
    let i = iterator.i;
    const array = iterator.array;
    for (; i < array.length; i++) {
      const el = array[i];
      if (Array.isArray(el)) {
        stack.push({ i: i + 1, array });
        stack.push({ i: 0, array: el });
        break;
      }
      if (typeof el === 'number') {
        if (expandRegisteredStyles) {
          hasResult = true;
          assignStyle(result, getStyle(el));
        }
      } else if (el) {
        hasResult = true;
        assignStyle(result, el);
      }
    }
  }
  return hasResult ? result : undefined;
};

const flattenClassNames = (input, flag) => {
  // optimize for this (common) scenario of input not needing flattening
  if (!Array.isArray(input)) {
    if (typeof input === 'number') {
      return getClassNames(input, 0);
    }
    return null;
  }
  let result = '';
  let position = 0;
  let hasSeenObject = false;
  const stack = [{ i: 0, array: input }];
  while (stack.length) {
    const iterator = stack.pop();
    let i = iterator.i;
    const array = iterator.array;
    for (; i < array.length; i++) {
      const el = array[i];
      if (Array.isArray(el)) {
        stack.push({ i: i + 1, array });
        stack.push({ i: 0, array: el });
        break;
      }
      if (typeof el === 'number') {
        result += `${getClassNames(el, position)} `;
        if (hasSeenObject) {
          // TODO(lmr): warn to console in this case to let user know? or do we not care?
          flag.deopt = true;
        }
      } else if (el) {
        hasSeenObject = true;
      }
      position++;
    }
  }
  return result;
};

const resolve = (styles, extraClassName) => {
  const flag = { deopt: false };
  const classes = flattenClassNames(styles, flag);
  const style = flattenStyle(styles, flag.deopt);
  return {
    className: !classes || flag.deopt ? extraClassName : `${extraClassName} ${classes}`,
    style: style ? transformToWebStyle(style) : null,
  };
};

const returnCopy = (original, result) => {
  if (original === result || typeof original === 'number') {
    return { ...result };
  }
  return result;
};

init();

// TODO(lmr):
// figure out a good strategy for SSR here.

const absoluteFillObject = {
  position: 'absolute',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

module.exports = {
  hairlineWidth: getHairlineWidth(),
  absoluteFillObject,
  absoluteFill: registerStyle('absoluteFill', absoluteFillObject),
  create,
  // NOTE:
  // `flatten` is exported separately from `resolve` because it mimics the RN api more closely
  // than `resolve`.
  flatten: style => returnCopy(style, flattenStyle(style, true)) || {},

  // TODO(lmr): should this be an internal API or something that we expose?
  resolve,

  reset,

  getStyleSheetHtml: injector.getStyleSheetHtml,
};
