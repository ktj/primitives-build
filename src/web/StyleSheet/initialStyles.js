const generateCss = require('./generateCss');
const injector = require('./injector');

function init() {
  const initialRules = {
    // https://github.com/facebook/css-layout#default-values
    rp_View: {
      alignItems: 'stretch',
      borderWidth: 0,
      borderStyle: 'solid',
      boxSizing: 'border-box',
      JsDisplay: 'flex',
      display: 'flex',
      flexBasis: 'auto',
      flexDirection: 'column',
      flexShrink: 0,
      margin: 0,
      padding: 0,
      position: 'relative',
      // button and anchor reset
      backgroundColor: 'transparent',
      color: 'inherit',
      font: 'inherit',
      textAlign: 'inherit',
      textDecorationLine: 'none',
      // list reset
      listStyle: 'none',
      // fix flexbox bugs
      minHeight: 0,
      minWidth: 0,
    },
    rp_ViewReset: {
      flexShrink: 0,
    },
    rp_Text: {
      color: 'inherit',
      display: 'inline',
      font: 'inherit',
      margin: 0,
      padding: 0,
      textDecorationLine: 'none',
      wordWrap: 'break-word',
      whiteSpace: 'pre-wrap',
    },
    rp_Image: {
      borderWidth: 0,
      height: 'auto',
      maxHeight: '100%',
      maxWidth: '100%',
      opacity: 0,
    },
  };

  Object.keys(initialRules).forEach(key => {
    const cssBody = generateCss(initialRules[key]);
    const css = `.${key}{${cssBody}}`;
    injector.addRule(key, css);
  });
}

module.exports = init;
