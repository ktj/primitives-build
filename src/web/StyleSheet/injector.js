const asap = require('asap');

const hasOwnProperty = Object.prototype.hasOwnProperty;

// TODO(lmr): to enable multiple bundles, we *could* just try and have registry be a global
let registry = {};
let isDirty = false;
let styleNode = null;

const getStyleText = () => {
  /* eslint prefer-template:0 */
  let result = '\n';
  for (const key in registry) {
    if (hasOwnProperty.call(registry, key)) {
      result += registry[key] + '\n';
    }
  }
  return result;
};

const frame = () => {
  if (!isDirty) return;
  if (!global.document) return;
  const document = global.document;
  isDirty = false;

  // TODO(lmr):
  // in the case of server-side rendering, we might want to include a "checksum" or
  // something to the end of the stylesheet so that we can ensure that we are re-initializing the
  // right one?  I'm not exactly sure how this would work though. The user might have to ensure
  // that there are not more than one "react-primitive" instances running on the page at once.
  if (!styleNode && document.querySelector) {
    styleNode = document.querySelector('style[data-react-primitives-stylesheet]');
  }

  if (!styleNode) {
    // Taken from
    // http://stackoverflow.com/questions/524696/how-to-create-a-style-tag-with-javascript
    const head = document.head || document.getElementsByTagName('head')[0];
    styleNode = document.createElement('style');

    styleNode.type = 'text/css';
    styleNode.setAttribute('data-react-primitives-stylesheet', '');
    head.appendChild(styleNode);
  }

  const css = getStyleText();

  if (styleNode.styleSheet) {
    styleNode.styleSheet.cssText = css;
  } else {
    /* eslint no-cond-assign:0 */
    let last;
    while (last = styleNode.lastChild) {
      styleNode.removeChild(last);
    }
    styleNode.appendChild(document.createTextNode(css));
  }
};

const addRule = (key, rule) => {
  registry[key] = rule;
  if (!isDirty) {
    isDirty = true;
    if (global.document) {
      // NOTE(lmr):
      // we are only calling this in a browser environment. On the server we can just use the
      // `getStyleSheetHtml` API.
      asap(frame);
    }
  }
};

const getStyleSheetHtml = () => `<style data-react-primitives-stylesheet>${getStyleText()}</style>`;

module.exports = {
  addRule,
  getStyleSheetHtml,
  reset: () => { registry = {}; },
};
