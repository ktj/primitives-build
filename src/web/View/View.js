const React = require('react');
const PropTypes = require('prop-types');
const StyleSheet = require('../StyleSheet');
const StyleSheetPropType = require('../propTypes/StyleSheetPropType');
const ViewStylePropTypes = require('./ViewStylePropTypes');
const normalizeNativeEvent = require('../Touchable/normalizeNativeEvent');
const applyPrimitiveMethods = require('../util/applyPrimitiveMethods');
const applyOnLayoutHandling = require('../util/applyOnLayoutHandling');
const Accessibility = require('../propTypes/Accessibility');
const { FLEXBOX_SUPPORTED, applyFlexboxPolyfill } = require('../util/flexboxSupport');

const { func, string } = PropTypes;

const propTypes = {
  accessibilityLabel: string,
  accessibilityLiveRegion: Accessibility.LiveRegionPropType,
  accessibilityRole: Accessibility.RolePropType,
  accessible: PropTypes.bool,
  children: PropTypes.node,
  onClick: func,
  onClickCapture: func,
  onMoveShouldSetResponder: func,
  onMoveShouldSetResponderCapture: func,
  onResponderGrant: func,
  onResponderMove: func,
  onResponderReject: func,
  onResponderRelease: func,
  onResponderTerminate: func,
  onResponderTerminationRequest: func,
  onStartShouldSetResponder: func,
  onStartShouldSetResponderCapture: func,
  onTouchCancel: func,
  onTouchCancelCapture: func,
  onTouchEnd: func,
  onTouchEndCapture: func,
  onTouchMove: func,
  onTouchMoveCapture: func,
  onTouchStart: func,
  onTouchStartCapture: func,
  pointerEvents: PropTypes.oneOf([
    'auto',
    'box-none',
    'box-only',
    'none',
  ]),
  style: StyleSheetPropType(ViewStylePropTypes),
  testID: string,
};

/**
 * React Native expects `pageX` and `pageY` to be on the `nativeEvent`, but
 * React doesn't include them for touch events.
 */
function normalizeHandler(handler) {
  if (!handler) {
    return undefined;
  }
  return (e) => {
    /* eslint no-param-reassign: 0 */
    if (e.nativeEvent.pageX === undefined) {
      e.nativeEvent = normalizeNativeEvent(e.nativeEvent);
    }
    handler(e);
  };
}

const VIEW_CLASSNAME = 'rp_View';

class View extends React.Component {
  constructor(props) {
    super(props);
    if (!FLEXBOX_SUPPORTED) {
      this.__setEl = this.__setEl.bind(this);
    }
  }
}

View.prototype.render = function render() {
  // TODO(lmr):
  // we should probably ensure that children is not a string, and enforce that those types of
  // children be rendered in a `<Text>` component
  const {
    pointerEvents,
    style,
    accessibilityRole,
  } = this.props;

  const passedStyle = !pointerEvents ? style : [style, { pointerEvents }];

  const resolvedStyle = StyleSheet.resolve(passedStyle, VIEW_CLASSNAME);
  const Component = (accessibilityRole && Accessibility.Components[accessibilityRole]) || 'div';
  const props = {
    className: resolvedStyle.className,
    style: resolvedStyle.style,
    children: this.props.children,
  };

  if (
    this.props.onClick ||
    this.props.onResponderGrant ||
    this.props.onTouchStart
  ) {
    Object.assign(props, {
      onClick: this.props.onClick,
      onClickCapture: this.props.onClickCapture,
      onMoveShouldSetResponder: this.props.onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture: this.props.onMoveShouldSetResponderCapture,
      onResponderGrant: this.props.onResponderGrant,
      onResponderMove: this.props.onResponderMove,
      onResponderReject: this.props.onResponderReject,
      onResponderRelease: this.props.onResponderRelease,
      onResponderTerminate: this.props.onResponderTerminate,
      onResponderTerminationRequest: this.props.onResponderTerminationRequest,
      onStartShouldSetResponder: this.props.onStartShouldSetResponder,
      onStartShouldSetResponderCapture: this.props.onStartShouldSetResponderCapture,
      onScroll: this.props.onScroll,
      onTouchCancel: normalizeHandler(this.props.onTouchCancel),
      onTouchCancelCapture: normalizeHandler(this.props.onTouchCancelCapture),
      onTouchEnd: normalizeHandler(this.props.onTouchEnd),
      onTouchEndCapture: normalizeHandler(this.props.onTouchEndCapture),
      onTouchMove: normalizeHandler(this.props.onTouchMove),
      onTouchMoveCapture: normalizeHandler(this.props.onTouchMoveCapture),
      onTouchStart: normalizeHandler(this.props.onTouchStart),
      onTouchStartCapture: normalizeHandler(this.props.onTouchStartCapture),
    });
  }

  if (
    this.props.accessible !== undefined ||
    this.props.accessibilityLabel ||
    this.props.accessibilityRole
  ) {
    Object.assign(props, {
      'aria-hidden': this.props.accessible === false,
      'aria-label': this.props.accessibilityLabel,
      'aria-live': this.props.accessibilityLiveRegion,
      role: accessibilityRole,
    });
  }

  if (this.props.onLayout) {
    this.applyOnLayoutIfNeeded();
  }

  if (!FLEXBOX_SUPPORTED) {
    // this is a performance optimization... and an ugly one at that.
    // we basically want to ensure that `StyleSheet.resolve` doesn't
    // have to get called twice for every update of every primitive,
    // so we cache this value here.
    props.ref = this.__setEl;
    this._lastResolvedStyle = resolvedStyle.style;
  }

  return React.createElement(Component, props);
};

View.propTypes = propTypes;
View.displayName = 'View';

applyPrimitiveMethods(View, VIEW_CLASSNAME);
applyFlexboxPolyfill(View);
applyOnLayoutHandling(View);

module.exports = View;
