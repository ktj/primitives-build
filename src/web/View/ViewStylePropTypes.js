const PropTypes = require('prop-types');
const BorderPropTypes = require('../propTypes/BorderPropTypes');
const ColorPropType = require('../propTypes/ColorPropType');
const LayoutPropTypes = require('../propTypes/LayoutPropTypes');
const TransformPropTypes = require('../propTypes/TransformPropTypes');
const ShadowPropTypes = require('../propTypes/ShadowPropTypes');

const { number, oneOf, string } = PropTypes;
const autoOrHiddenOrVisible = oneOf(['auto', 'hidden', 'visible', 'scroll']);
const hiddenOrVisible = oneOf(['hidden', 'visible']);

module.exports = {
  ...BorderPropTypes,
  ...LayoutPropTypes,
  ...TransformPropTypes,
  ...ShadowPropTypes,
  // RN-only styles
  elevation: number,

  backfaceVisibility: hiddenOrVisible,
  backgroundColor: ColorPropType,
  opacity: number,
  overflow: autoOrHiddenOrVisible,
  // TODO(lmr): we probably want to shim the RN stylesheet or something to strip
  // these styles, or else it will trigger a redbox in RN.
  /* @platform web */
  WebkitOverflowScrolling: string,
  backgroundAttachment: string,
  backgroundClip: string,
  backgroundImage: string,
  backgroundPosition: string,
  backgroundOrigin: oneOf([
    'border-box',
    'content-box',
    'padding-box',
  ]),
  backgroundRepeat: string,
  backgroundSize: string,
  cursor: string,
  outline: string,
  overflowX: autoOrHiddenOrVisible,
  overflowY: autoOrHiddenOrVisible,
  userSelect: string,
  visibility: hiddenOrVisible,
  zIndex: number,
};
