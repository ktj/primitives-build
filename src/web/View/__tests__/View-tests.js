import React from 'react';
import renderer from 'react-test-renderer';
import View from '../View';
import StyleSheet from '../../StyleSheet';

describe('<View />', () => {
  it('empty with no children', () => {
    expect(renderer.create(
      <View />
    )).toMatchSnapshot();
  });

  it('empty with children', () => {
    expect(renderer.create(
      <View>
        <View />
      </View>
    )).toMatchSnapshot();
  });

  it('with single registered style', () => {
    const styles = StyleSheet.create({
      foo: {
        width: 20,
        height: 20,
      },
    });
    expect(renderer.create(
      <View
        style={styles.foo}
      />
    )).toMatchSnapshot();
  });

  it('with single inline style', () => {
    expect(renderer.create(
      <View
        style={{
          width: 20,
          height: 20,
        }}
      >
        Hello World!
      </View>
    )).toMatchSnapshot();
  });

  it('with multiple styles and some inline', () => {
    const styles = StyleSheet.create({
      foo: {
        width: 20,
        height: 20,
      },
      bar: {
        backgroundColor: 'red',
      },
    });
    expect(renderer.create(
      <View
        style={[
          styles.foo,
          styles.bar,
          { opacity: 0.5 },
        ]}
      />
    )).toMatchSnapshot();
  });

  it('with a11y props', () => {
    expect(renderer.create(
      <View
        style={{
          width: 20,
          height: 20,
        }}
        accessibilityLabel="Label"
        accessible
      >
        Hello World!
      </View>
    )).toMatchSnapshot();
  });

  it('with a11y role', () => {
    expect(renderer.create(
      <View
        style={{
          width: 20,
          height: 20,
        }}
        accessibilityRole="button"
      >
        Hello World!
      </View>
    )).toMatchSnapshot();
  });
});
