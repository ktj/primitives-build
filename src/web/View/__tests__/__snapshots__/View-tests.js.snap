exports[`<View /> empty with children 1`] = `
<div
  className="rp_View"
  style={null}>
  <div
    className="rp_View"
    style={null} />
</div>
`;

exports[`<View /> empty with no children 1`] = `
<div
  className="rp_View"
  style={null} />
`;

exports[`<View /> with a11y props 1`] = `
<div
  aria-hidden={false}
  aria-label="Label"
  aria-live={undefined}
  className="rp_View"
  role={undefined}
  style={
    Object {
      "height": "20px",
      "width": "20px",
    }
  }>
  Hello World!
</div>
`;

exports[`<View /> with a11y role 1`] = `
<button
  aria-hidden={false}
  aria-label={undefined}
  aria-live={undefined}
  className="rp_View"
  role="button"
  style={
    Object {
      "height": "20px",
      "width": "20px",
    }
  }>
  Hello World!
</button>
`;

exports[`<View /> with multiple styles and some inline 1`] = `
<div
  className="rp_View r1pr8dlo rsp2czt1 "
  style={
    Object {
      "opacity": 0.5,
    }
  } />
`;

exports[`<View /> with single inline style 1`] = `
<div
  className="rp_View"
  style={
    Object {
      "height": "20px",
      "width": "20px",
    }
  }>
  Hello World!
</div>
`;

exports[`<View /> with single registered style 1`] = `
<div
  className="rp_View r1pr8dlo"
  style={null} />
`;
